man/safe_mysqld.1
.TH safe_mysqld 1 "19 December 2000" "MySQL 3.23" "MySQL database"
.SH NAME
safe_mysqld \- start the mysqld daemon on Unix.
.SH SYNOPSIS
.B safe_mysqld
.RB [ \-\-basedir=\fP\fIpath\fP ]
.RB [ \-\-core\-file\-size=# ]
.RB [ \-\-defaults\-extra\-file=\fP\fIpath\fP ]
.RB [ \-\-defaults\-file=\fP\fIpath\fP ]
.RB [ \-\-open\-files=# ]
.RB [ \-\-datadir=\fP\fIpath\fP ]
.RB [ \-\-err\-log=\fP\fIpath \fP]
.RB [ \-\-ledir=path ]
.RB [ \-\-log=\fP\fIpath\fP ]
.RB [ \-\-no\-defaults ]
.RB [ \-\-open\-files=# ]
.RB [ \-\-pid\-file=\fP\fIpath\fP ]
.RB [ \-\-port=# ]
.RB [ \-\-socket=\fP\fIpath\fP ]
.RB [ \-\-timezone=# ]
.RB [ \-\-user=# ]
.SH DESCRIPTION
safe_mysqld adds some safety features such as restarting the server when an
error occurs and logging run-time information to a log file.
.BR
.TP 
.BR \-\-basedir=\fP\fIpath \fP
.TP 
.BR \-\-core\-file\-size=# 
.TP 
.BR \-\-defaults\-extra\-file=\fP\fIpath \fP
.TP 
.BR \-\-defaults\-file=\fP\fIpath \fP
.TP 
.BR \-\-open\-files=# 
Size of the core file mysqld should be able to create. Passed to ulimit \-c. 
.TP 
.BR \-\-datadir=\fP\fIpath \fP
.TP 
.BR \-\-err\-log=\fP\fIpath \fP
.TP 
.BR \-\-ledir=\fP\fIpath \fP
Path to mysqld 
.TP 
.BR \-\-log=\fP\fIpath \fP
.TP 
.BR \-\-no\-defaults 
.TP 
.BR \-\-open\-files=# 
Number of files mysqld should be able to open. Passed to ulimit \-n. 
.TP 
.BR \-\-pid\-file=\fP\fIpath \fP
.TP 
.BR \-\-port=# 
.TP 
.BR \-\-socket=\fP\fIpath \fP
.TP 
.BR \-\-timezone=# 
Set the timezone (the TZ) variable to the value of this parameter. 
.TP 
.BR \-\-user=# 
.SH NOTE
Note that all options on the command line to safe_mysqld are passed to mysqld. If you wants to use any options in safe_mysqld that mysqld doesn't support, you must specify these in the option file. 
.SH "SEE ALSO"
isamchk (1),
isamlog (1),
mysqlaccess (1),
mysqladmin (1),
mysqlbug (1),
mysqld (1),
mysqldump (1),
mysqlshow (1),
msql2mysql (1),
perror (1),
replace (1),
safe_mysqld (1),
which1 (1),
zap (1),
.SH AUTHOR
Ver 1.0, distribution 3.23.29a
Michael (Monty) Widenius (monty@tcx.se),
TCX Datakonsult AB (http://www.tcx.se).
This software comes with no warranty.
Manual page by L. (Kill-9) Pedersen 
(kill-9@kill\-9.dk), Mercurmedia Data Model Architect /
system developer (http://www.mercurmedia.com)

.\" end of man page


ChangeLog
2008-10-22	The InnoDB Team

	* dict/dict0dict.c, dict/dict0mem.c, handler/ha_innodb.cc,
	handler/ha_innodb.h, include/dict0dict.h, include/dict0mem.h,
	row/row0mysql.c:
	Fix Bug#39830 Table autoinc value not updated on first insert
	Fix Bug#35498 Cannot get table test/table1 auto-inccounter value in
	::info
	Fix Bug#36411 "Failed to read auto-increment value from storage
	engine" in 5.1.24 auto-inc

2008-10-22	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Fix Bug#40224 New AUTOINC changes mask reporting of deadlock/timeout
	errors

2008-10-09	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#39939 DROP TABLE/DISCARD TABLESPACE takes long time in
	buf_LRU_invalidate_tablespace()

2008-10-08	The InnoDB Team

	* dict/dict0crea.c, trx/trx0roll.c, include/row0mysql.h,
	row/row0merge.c, row/row0mysql.c: When dropping a table, hold the
	data dictionary latch until the transaction has been committed.
	The data dictionary latch is supposed to prevent lock waits and
	deadlocks in the data dictionary tables.  Due to this bug,
	DROP TABLE could cause a deadlock or hang.  Note that because of
	Bug#33650 and Bug#39833, MySQL may also drop a (temporary) table
	when executing CREATE INDEX or ALTER TABLE ... ADD INDEX.

2008-10-04	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug39438-master.opt,
	mysql-test/innodb_bug39438.result, mysql-test/innodb_bug39438.test:
	Fix Bug#39438 Testcase for Bug#39436 crashes on 5.1 in
	fil_space_get_latch

2008-10-04	The InnoDB Team

	* include/lock0lock.h, lock/lock0lock.c,
	mysql-test/innodb_bug38231.result, mysql-test/innodb_bug38231.test,
	row/row0mysql.c:
	Fix Bug#38231 Innodb crash in lock_reset_all_on_table() on TRUNCATE +
	LOCK / UNLOCK

2008-10-04	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#35498 Cannot get table test/table1 auto-inccounter value in
	::info

2008-10-04	The InnoDB Team

	* handler/ha_innodb.cc, handler/ha_innodb.h:
	Fix Bug#37788 InnoDB Plugin: AUTO_INCREMENT wrong for compressed
	tables

2008-10-04	The InnoDB Team

	* dict/dict0dict.c, handler/ha_innodb.cc, handler/ha_innodb.h,
	include/dict0dict.h, include/dict0mem.h, row/row0mysql.c:
	Fix Bug#39830 Table autoinc value not updated on first insert

2008-10-03	The InnoDB Team

	* mysql-test/innodb-index.test, mysql-test/innodb-index.result,
	mysql-test/innodb-timeout.test, mysql-test/innodb-timeout.result,
	srv/srv0srv.c, include/srv0srv.h,
	handler/ha_innodb.cc, include/ha_prototypes.h:
	Fix Bug#36285 innodb_lock_wait_timeout is not dynamic, not per session

2008-09-19	The InnoDB Team

	* os/os0proc.c:
	Fix a memory leak on Windows. The memory leak was due to wrong
	parameters passed into VirtualFree() call. As the result, the
	call fails with Windows error 87.

2008-09-17	The InnoDB Team

	* mysql-test/innodb.result, mysql-test/innodb-zip.result,
	mysql-test/innodb-zip.test, mysql-test/innodb.test,
	ibuf/ibuf0ibuf.c, dict/dict0crea.c, dict/dict0load.c, dict/dict0boot.c,
	include/dict0dict.h, include/trx0trx.h,
	dict/dict0dict.c, trx/trx0trx.c,
	include/ha_prototypes.h, handler/ha_innodb.cc:
	When creating an index in innodb_strict_mode, check that the
	maximum record size will never exceed the B-tree page size limit.
	For uncompressed tables, there should always be enough space for
	two records in an empty B-tree page.  For compressed tables, there
	should be enough space for storing two node pointer records or one
	data record in an empty page in uncompressed format.
	The purpose of this check is to guarantee that INSERT or UPDATE
	will never fail due to too big record size.

2008-09-17	The InnoDB Team

	* btr/btr0cur.c, data/data0data.c, include/page0zip.h,
	include/page0zip.ic, page/page0zip.c,
	mysql-test/innodb_bug36172.test:
	Prevent infinite B-tree page splits in compressed tables by
	ensuring that there will always be enough space for two node
	pointer records in an empty B-tree page.  Also, require that at
	least one data record will fit in an empty compressed page.  This
	will reduce the maximum size of records in compressed tables.

2008-09-09	The InnoDB Team

	* mysql-test/innodb.result:
	Fix the failing innodb test by merging changes that MySQL made to
	that file (r2646.12.1 in MySQL BZR repository)

2008-09-09	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc.test:
	Fix Bug#38839 auto increment does not work properly with InnoDB after
	update

2008-09-09	The InnoDB Team

	* dict/dict0dict.c, handler/handler0alter.cc, include/dict0dict.h,
	mysql-test/innodb-index.result, mysql-test/innodb-index.test:
	Fix Bug#38786 InnoDB plugin crashes on drop table/create table with FK

2008-08-21	The InnoDB Team

	* handler/ha_innodb.cc, include/ha_prototypes.h, row/row0sel.c:
	Fix Bug#37885 row_search_for_mysql may gap lock unnecessarily with SQL
	comments in query

2008-08-21	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#38185 ha_innobase::info can hold locks even when called with
	HA_STATUS_NO_LOCK

2008-08-18	The InnoDB Team

	* buf/buf0buf.c, buf/buf0lru.c, include/buf0buf.ic, include/univ.i:
	Introduce UNIV_LRU_DEBUG for debugging the LRU buffer pool cache

2008-08-08	The InnoDB Team

	* buf/buf0lru.c, include/buf0buf.h:
	Fix two recovery bugs that could lead to a crash in debug builds with
	small buffer size

2008-08-07	The InnoDB Team

	* btr/btr0cur.c, handler/ha_innodb.cc, include/srv0srv.h,
	srv/srv0srv.c:
	Add a parameter innodb_stats_sample_pages to allow users to control
	the number of index dives when InnoDB estimates the cardinality of
	an index (ANALYZE TABLE, SHOW TABLE STATUS etc)

2008-08-07	The InnoDB Team

	* trx/trx0i_s.c:
	Fix a bug that would lead to a crash if a SELECT was issued from the
	INFORMATION_SCHEMA tables and there are rolling back transactions at
	the same time

2008-08-06	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, ibuf/ibuf0ibuf.c, include/btr0cur.h,
	include/trx0roll.h, include/trx0types.h, row/row0purge.c,

scripts/safe_mysqld.sh
#!/bin/sh
# Copyright Abandoned 1996 TCX DataKonsult AB & Monty Program KB & Detron HB
# This file is public domain and comes with NO WARRANTY of any kind
#
# scripts to start the MySQL daemon and restart it if it dies unexpectedly
#
# This should be executed in the MySQL base directory if you are using a
# binary installation that has other paths than you are using.
#
# mysql.server works by first doing a cd to the base directory and from there
# executing safe_mysqld

trap '' 1 2 3 15			# we shouldn't let anyone kill us

defaults=
case "$1" in
    --no-defaults|--defaults-file=*|--defaults-extra-file=*)
      defaults="$1"; shift
      ;;
esac

parse_arguments() {
  # We only need to pass arguments through to the server if we don't
  # handle them here.  So, we collect unrecognized options (passed on
  # the command line) into the args variable.
  pick_args=$1; shift

  for arg do
    case "$arg" in
      # these get passed explicitly to mysqld
      --basedir=*) MY_BASEDIR_VERSION=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --datadir=*) DATADIR=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --pid-file=*) pid_file=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --user=*)    user=`echo "$arg" | sed -e "s;--[^=]*=;;"` ; SET_USER=1 ;;

      # these two might have been set in a [safe_mysqld] section of my.cnf
      # they get passed via environment variables to safe_mysqld
      --socket=*)  MYSQL_UNIX_PORT=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --port=*)    MYSQL_TCP_PORT=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;

      # safe_mysqld-specific options
      --ledir=*)   ledir=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --err-log=*) err_log=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      # QQ The --open-files should be removed
      --open-files=*) open_files=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --open-files-limit=*) open_files=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --core-file-size=*) core_file_size=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --timezone=*) TZ=`echo "$arg" | sed -e "s;--[^=]*=;;"` ; export TZ; ;;
      --mysqld=*)   MYSQLD=`echo "$arg" | sed -e "s;--[^=]*=;;"` ;;
      --mysqld-version=*)
	tmp=`echo "$arg" | sed -e "s;--[^=]*=;;"`
	if test -n "$tmp"
	then
	  MYSQLD="mysqld-$tmp"
	else
	  MYSQLD="mysqld"
	fi
	;;
      *)
        if test $pick_args -eq 1
        then
          # This sed command makes sure that any special chars are quoted,
          # so the arg gets passed exactly to the server.
          args="$args "`echo "$arg" | sed -e 's,\([^a-zA-Z0-9_.-]\),\\\\\1,g'`
        fi
        ;;
    esac
  done
}

MY_PWD=`pwd`
# Check if we are starting this relative (for the binary release)
if test -d $MY_PWD/data/mysql -a -f ./share/mysql/english/errmsg.sys -a \
 -x ./bin/mysqld
then
  MY_BASEDIR_VERSION=$MY_PWD		# Where bin, share and data are
  ledir=$MY_BASEDIR_VERSION/bin		# Where mysqld is
  DATADIR=$MY_BASEDIR_VERSION/data
  if test -z "$defaults"
  then
    defaults="--defaults-extra-file=$MY_BASEDIR_VERSION/data/my.cnf"
  fi
# Check if this is a 'moved install directory'
elif test -f ./var/mysql/db.frm -a -f ./share/mysql/english/errmsg.sys -a \
 -x ./libexec/mysqld
then
  MY_BASEDIR_VERSION=$MY_PWD		# Where libexec, share and var are
  ledir=$MY_BASEDIR_VERSION/libexec	# Where mysqld is
  DATADIR=$MY_BASEDIR_VERSION/var
else
  MY_BASEDIR_VERSION=@prefix@
  DATADIR=@localstatedir@
  ledir=@libexecdir@
fi

MYSQL_UNIX_PORT=${MYSQL_UNIX_PORT:-@MYSQL_UNIX_ADDR@}
MYSQL_TCP_PORT=${MYSQL_TCP_PORT:-@MYSQL_TCP_PORT@}
user=@MYSQLD_USER@

# Use the mysqld-max binary by default if the user doesn't specify a binary
if test -x $ledir/mysqld-max
then
  MYSQLD=mysqld-max
else
  MYSQLD=mysqld
fi

# these rely on $DATADIR by default, so we'll set them later on
pid_file=
err_log=
SET_USER=0

# Get first arguments from the my.cnf file, groups [mysqld] and [safe_mysqld]
# and then merge with the command line arguments
if test -x ./bin/my_print_defaults
then
  print_defaults="./bin/my_print_defaults"
elif test -x @bindir@/my_print_defaults
then
  print_defaults="@bindir@/my_print_defaults"
elif test -x @bindir@/mysql_print_defaults
then
  print_defaults="@bindir@/mysql_print_defaults"
else
  print_defaults="my_print_defaults"
fi

args=
parse_arguments 0 `$print_defaults $defaults mysqld server safe_mysqld`
parse_arguments 1 "$@"

if test ! -x $ledir/$MYSQLD
then
  echo "The file $ledir/$MYSQLD doesn't exist or is not executable"
  echo "Please do a cd to the mysql installation directory and restart"
  echo "this script from there as follows:"
  echo "./bin/safe_mysqld".
  exit 1
fi

if test -z "$pid_file"
then
  pid_file=$DATADIR/`@HOSTNAME@`.pid
else
  case "$pid_file" in
    /* ) ;;
    * )  pid_file="$DATADIR/$pid_file" ;;
  esac
fi
test -z "$err_log"  && err_log=$DATADIR/`@HOSTNAME@`.err

export MYSQL_UNIX_PORT
export MYSQL_TCP_PORT


NOHUP_NICENESS="nohup"
if test -w /
then
  NOHUP_NICENESS=`nohup nice 2>&1`
 if test $? -eq 0 && test x"$NOHUP_NICENESS" != x0 && nice --1 echo foo > /dev/null 2>&1
 then
    NOHUP_NICENESS="nice -n $NOHUP_NICENESS nohup"
  else
    NOHUP_NICENESS="nohup"
  fi
fi

USER_OPTION=""
if test -w /
then
  if test "$user" != "root" -o $SET_USER = 1
  then
    USER_OPTION="--user=$user"
  fi
  # If we are root, change the err log to the right user.
  touch $err_log; chown $user $err_log
  if test -n "$open_files"
  then
    ulimit -n $open_files
  fi
  if test -n "$core_file_size"
  then
    ulimit -c $core_file_size
  fi
fi

#
# If there exists an old pid file, check if the daemon is already running
# Note: The switches to 'ps' may depend on your operating system
if test -f $pid_file
then
  PID=`cat $pid_file`
  if @CHECK_PID@
  then
    if @FIND_PROC@
    then    # The pid contains a mysqld process
      echo "A mysqld process already exists"
      echo "A mysqld process already exists at " `date` >> $err_log
      exit 1
    fi
  fi
  rm -f $pid_file
  if test -f $pid_file
  then
    echo "Fatal error: Can't remove the pid file: $pid_file"
    echo "Fatal error: Can't remove the pid file: $pid_file at " `date` >> $err_log
    echo "Please remove it manually and start $0 again"
    echo "mysqld daemon not started"
    exit 1
  fi
fi

#
# Uncomment the following lines if you want all tables to be automaticly
# checked and repaired at start
#
# echo "Checking tables in $DATADIR"
# $MY_BASEDIR_VERSION/bin/myisamchk --silent --force --fast --medium-check -O key_buffer=64M -O sort_buffer=64M $DATADIR/*/*.MYI
# $MY_BASEDIR_VERSION/bin/isamchk --silent --force -O sort_buffer=64M $DATADIR/*/*.ISM

echo "Starting $MYSQLD daemon with databases from $DATADIR"

# Does this work on all systems?
#if type ulimit | grep "shell builtin" > /dev/null
#then
#  ulimit -n 256 > /dev/null 2>&1		# Fix for BSD and FreeBSD systems
#fi

echo "`date +'%y%m%d %H:%M:%S  mysqld started'`" >> $err_log
while true
do
  rm -f $MYSQL_UNIX_PORT $pid_file	# Some extra safety
  if test -z "$args"
  then
    $NOHUP_NICENESS $ledir/$MYSQLD $defaults --basedir=$MY_BASEDIR_VERSION --datadir=$DATADIR $USER_OPTION --pid-file=$pid_file @MYSQLD_DEFAULT_SWITCHES@ >> $err_log 2>&1
  else
    eval "$NOHUP_NICENESS $ledir/$MYSQLD $defaults --basedir=$MY_BASEDIR_VERSION --datadir=$DATADIR $USER_OPTION --pid-file=$pid_file @MYSQLD_DEFAULT_SWITCHES@ $args >> $err_log 2>&1"
  fi
  if test ! -f $pid_file		# This is removed if normal shutdown
  then
    break
  fi

  if @IS_LINUX@
  then
    # Test if one process was hanging.
    # This is only a fix for Linux (running as base 3 mysqld processes)
    # but should work for the rest of the servers.
    # The only thing is ps x => redhat 5 gives warnings when using ps -x.
    # kill -9 is used or the process won't react on the kill.
    numofproces=`ps xa | grep -v "grep" | grep -c $ledir/$MYSQLD`
    echo -e "\nNumber of processes running now: $numofproces" | tee -a $err_log
    I=1
    while test "$I" -le "$numofproces"
    do 
      PROC=`ps xa | grep $ledir/$MYSQLD | grep -v "grep" | sed -n '$p'` 
	for T in $PROC
	do
	  break
	done
	#    echo "TEST $I - $T **"
	if kill -9 $T
	then
	  echo "$MYSQLD process hanging, pid $T - killed" | tee -a $err_log
	else 
	  break
	fi
	I=`expr $I + 1`
    done
  fi

  echo "`date +'%y%m%d %H:%M:%S'`  mysqld restarted" | tee -a $err_log
done

echo "`date +'%y%m%d %H:%M:%S'`  mysqld ended" | tee -a $err_log
echo "" | tee -a $err_log
